### DISCUSSION TITLE
Solved: File Menu event appears with blank Alias?

### DISCUSSION THREAD

**Discussion by breaker-of-stone** [Taitava/obsidian-shellcommands#313] [Debugging and testing]
([dup. from Obsidian forum](https://forum.obsidian.md/t/shell-commands-plugin/23497/29))

Issue, or lack of understanding?:  I am using Obs. 1.1.9 on a Mac Pro under MacOS Ventura.

I have one shell command registered and **it functions correctly**, however where the alias should appear in the file context menu is a blank item:

![image](https://user-images.githubusercontent.com/8778605/210479582-8b6341aa-e979-4177-9356-7adf7860e49c.png)

What am I missing?
A screen shot of the settings summary:

![image](https://user-images.githubusercontent.com/8778605/210479609-272664a0-77fc-4dfc-a6f0-ce5445d3d19f.png)

Command is:
open -a Atom {{event_file_path:absolute}}

(Obsidian) SYSTEM INFO:
Obsidian version: v1.1.9
Installer version: v1.1.8
Operating system: Darwin Kernel Version 22.1.0: Sun Oct 9 20:14:54 PDT 2022; root:xnu-8792.41.9~2/RELEASE_X86_64 22.1.0
Login status: not logged in
Insider build toggle: off
Live preview: on
Legacy editor: off
Base theme: dark
Community theme: none
Snippets enabled: 0
Restricted mode: off
Plugins installed: 3
Plugins enabled: 1
1: Shell commands v0.17.0

**Comment by Taitava** [comment]
Hi! Thank you for your post and especially for your detailed debug information! 🙂 

# I suspected a wrong thing

I've never seen this bug before, ~~but I think I found something that might lead me to a solution on how to fix it. It's clearly a bug in the plugin.~~

![kuva](https://user-images.githubusercontent.com/13002142/210586836-750d223e-0a76-4ef7-b3e0-2222e6f25af6.png)

~~I assume you have the setting _Preview variables in command palette and menus_ turned off? I usually have it turned on, but I tried to turn it off now and then I was able to reproduce the bug.~~

~~So, for a temporary workaround, can you please turn the setting on and see if your menu item becomes visible? I'll try to come up with a fix so that menu items would work also when the setting is turned off.~~

Actually, the problem I described happened in my development which had unreleased `0.18.0` code in place. One change of the future coming `0.18.0` had broken the menu in a way that when the said setting is turned off, the menu item disappears. I fixed it, but realised I only fixed a bug which never were in production. In other words, my fix won't fix the bug you are seeing 🙁.

So, the updated situation is that I haven't yet been able to reproduce your problem. I'll try to find a solution for this, after releasing `0.18.0`. But at least this issue led me notice another similar bug that might otherwise have been gone unnoticedly to production. 😅

**Reply by Taitava** [reply]
I'll also need to create a test so that this bug won't appear again after the fix.

**Reply by Taitava** [reply]
~~Will be fixed in `0.18.0`.~~

**Comment by Taitava** [comment]
# Another possible solution

I still can't say that I would have found the cause for this, but here's at least another thing that might help: I've made [a fix on 2023-01-01](https://github.com/Taitava/obsidian-shellcommands/commit/a586ab480b0048ce889e5b57f1af775821085981) (not published yet, will be published in `0.18.0`) that might affect the problem. However, that was a blind fix on a problem that I had seen in the code, but never seen happening in practise. So: I can't be certain that this will fix it, but we'll see.

So, after `0.18.0` is released, @breaker-of-stone please report here whether the problem goes away or if it's still present. 🙂

**Reply by breaker-of-stone** [reply]
Ok, keep my eyes on the horizon.  I'm guessing late Jan / early Feb?

**Reply by Taitava** [reply]
I hope I can release `0.18.0` today, but will see. I tried to release it last weekend, but noticed problems during my testing phase. I hope I've been able to fix them now.

**Reply by Taitava** [reply]
@breaker-of-stone `0.18.0` is now released.

**Comment by breaker-of-stone** [comment]
No change in behaviour with 0.18.0, I'm afraid.
Have a chat session (UCT + 0800) to validate my technique?

**Reply by Taitava** [reply]
Too bad it didn't help 🙁.

Yea, we can have a chat on Discord some day. My TimeZone is UTC+2 (Helsinki). My Discord user is Jarkko Linnanvirta #5429. I'm not very often online in Discord, but we can agree some time there. Just send me a message there.

I can also borrow my friend's Mac some day and try to see if I can reproduce the problem on macOS. But the problem sounds like if it would happen on all Macs, there would probably be multiple users reporting it.

**Reply by Taitava** [reply]
# Debug

One more thing that came to my mind that might help us hunt down the bug. If you want you can gather up some more debug information:

1. Open up `.obsidian/plugins/obsidian-shellcommands/data.json` in an external editor and change `"debug":` from `false` to `true`. This will enable extended information logging in Obsidian's console.
2. Reload the shell commands plugin's settings e.g. by simply restarting Obsidian.
3. Open up console by pressing `Cmd` + `Option` + `I`.
4. Perform a right click on some file in navigation to open up the file menu.
    - See if anything interesting comes up in the console log. The SC plugin might output quite a lot of stuff there.
    - If you see any **red errors** there, it's a clear indicator of a bug. (This kind of errors can be seen in the console even if `debug` is `false`).
    - If no red errors appear, the plugin's own loggings in the console can still be helpful. If you want, you can post the console log content here. Just make sure it doesn't contain any private information you don't want to show publicly.

**Reply by breaker-of-stone** [reply]
Followed steps above, Dev Console doesn't display _any_ logging info during menu actions or ShellCommand configuration actions.  Filter is set to All Levels.
Intermittent reflow messages while starting the app or opening ShellCommand config e.g:

> [Violation] Forced reflow while executing JavaScript took 34ms

See message from me on Discord.  Screen share after 1000 Pori / 0800 UCT?

**Reply by breaker-of-stone** [reply]
Added a second, identical shellcommand and observed that a second blank File menu entry appears under the first one.  It also functions as expected, opening the file in Atom.

No messages at all in the Developer Console - not even when disabling/ enabling the plugin.  Is that expected?

Happy to Discord with enough warning - message IanC#5816 when you can.

**Reply by breaker-of-stone** [reply]
I notice:
./plugins/obsidian-shellcommands/main.js:6121        debug: false,
... which may be a default, or it may override the data.json setting, explaining why we're not getting any debugging output.
??

**Reply by breaker-of-stone** [reply]
PBKAC!  Tactical error: had two vaults (windows) open and looking at Dev Console in the wrong vault, which wasn't configured for debug.

Switched to the debug-able vault and voila!  

The Alias details tab does not log any parsing of the title field, yet it is stored in data.json correctly and parsed out when opening the file context menu.  ??

Note command ID varies in transcript below, as I repeated the process several times to verify.

## Procedure
  * Remove ShellCommand plugin
  * Re-install ShellCommand
  * Create first ShellCommand:
**'New shell command'**:
```
plugin:obsidian-shellcommands:188 IDGenerator: Generated id p5abimkf0d
plugin:obsidian-shellcommands:188 IDGenerator: Adding id p5abimkf0d to the list of reserved ids.
plugin:obsidian-shellcommands:188 Registering shell command #p5abimkf0d...
plugin:obsidian-shellcommands:188 Registered.
plugin:obsidian-shellcommands:188 Create command field for command #p5abimkf0d (NEW)
plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse  with 41 variables.
plugin:obsidian-shellcommands:188 parseVariables(): Parsing content  succeeded.
plugin:obsidian-shellcommands:188 Created.
plugin:obsidian-shellcommands:188 New empty command created.
```
**Click Events tab of new command**:
```
plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse  with 41 variables.
plugin:obsidian-shellcommands:188 parseVariables(): Parsing content  succeeded.
plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse  with 41 variables.
plugin:obsidian-shellcommands:188 parseVariables(): Parsing content  succeeded.
plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse  with 41 variables.
plugin:obsidian-shellcommands:188 parseVariables(): Parsing content  succeeded.
app.js:1 [Violation] 'click' handler took 181ms
[Violation] Forced reflow while executing JavaScript took 70ms
```
**Turn on 'File Menu' event**:
\<no log messages\>

**Close Events tab, click in command field**:
\<no log messages\>

**Paste new command into field**:
```
plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse open -a Atom {{event_file_path:absolute}} with 41 variables.
plugin:obsidian-shellcommands:188 Variable_EventFilePath.getValue(): Caught a VariableError and will determine how to handle it: {{event_file_path}}: This variable can only be used during events: File menu, File created, File content modified, File deleted, File moved, File renamed
plugin:obsidian-shellcommands:188 Variable {{event_file_path}} is not available. Will prevent shell command execution and show visible error messages.
plugin:obsidian-shellcommands:188 parseVariables(): Parsing content open -a Atom {{event_file_path:absolute}} failed.
plugin:obsidian-shellcommands:188 Creating new command p5abimkf0d: open -a Atom {{event_file_path:absolute}}
plugin:obsidian-shellcommands:188 Command created.
```
**Click Alias tab of this command, enter an alias title**:
\<no log messages\> **!!**

**Right-clicking an individual file name**:
```plugin:obsidian-shellcommands:188 TShellCommand dm97vrm1hz: Getting non preactions dependent variables.
plugin:obsidian-shellcommands:188 TShellCommand dm97vrm1hz: Getting preactions dependent variables.
plugin:obsidian-shellcommands:188 TShellCommand dm97vrm1hz: Getting preactions.
plugin:obsidian-shellcommands:188 TShellCommand dm97vrm1hz: Getting preactions dependent variables.
plugin:obsidian-shellcommands:188 TShellCommand dm97vrm1hz: Getting preactions.
plugin:obsidian-shellcommands:188 Parsing process: Count variable sets: 2
plugin:obsidian-shellcommands:188 Parsing process: Count variables in current set: 41
plugin:obsidian-shellcommands:188 Starting to parse 'shell_command': open -a Atom {{event_file_path:absolute}}
plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse open -a Atom {{event_file_path:absolute}} with 41 variables.
plugin:obsidian-shellcommands:188 parseVariables(): Parsing content open -a Atom {{event_file_path:absolute}} succeeded.
plugin:obsidian-shellcommands:188 Starting to parse 'alias': Open in Atom
plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse Open in Atom with 41 variables.
plugin:obsidian-shellcommands:188 parseVariables(): Parsing content Open in Atom succeeded.
plugin:obsidian-shellcommands:188 Starting to parse 'environment_variable_path_augmentation': 
plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse  with 41 variables.
plugin:obsidian-shellcommands:188 parseVariables(): Parsing content  succeeded.
plugin:obsidian-shellcommands:188 Starting to parse 'stdinContent': undefined
plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse undefined with 41 variables.
plugin:obsidian-shellcommands:188 parseVariables(): Parsing content undefined succeeded.
plugin:obsidian-shellcommands:188 Starting to parse 'output_wrapper_stdout': undefined
plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse undefined with 41 variables.
plugin:obsidian-shellcommands:188 parseVariables(): Parsing content undefined succeeded.
plugin:obsidian-shellcommands:188 Starting to parse 'output_wrapper_stderr': undefined
plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse undefined with 41 variables.
plugin:obsidian-shellcommands:188 parseVariables(): Parsing content undefined succeeded.
```

Tranche 2 - **Clicking the blank menu option opens the file in Atom**:
```
plugin:obsidian-shellcommands:188 TShellCommand dm97vrm1hz: Getting preactions.
plugin:obsidian-shellcommands:188 Going to prepare possible Preactions with an already started variable parsing process.
plugin:obsidian-shellcommands:188 No Preactions to perform. This is ok.
plugin:obsidian-shellcommands:188 Parsing all the rest of the variables (if there are any left).
6plugin:obsidian-shellcommands:188 Previous parsing succeeded? true
plugin:obsidian-shellcommands:188 Parsing process: Count variables in current set: 0
plugin:obsidian-shellcommands:188 Continuing parsing 'shell_command': open -a Atom \/Users\/ian\/Desktop\/Info\/Projects\/NoiseBin\/Design\/Library\/tools\/log\-install\-czmq\-library\.md
plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse open -a Atom \/Users\/ian\/Desktop\/Info\/Projects\/NoiseBin\/Design\/Library\/tools\/log\-install\-czmq\-library\.md with 0 variables.
plugin:obsidian-shellcommands:188 parseVariables(): Parsing content open -a Atom \/Users\/ian\/Desktop\/Info\/Projects\/NoiseBin\/Design\/Library\/tools\/log\-install\-czmq\-library\.md succeeded.
plugin:obsidian-shellcommands:188 Continuing parsing 'alias': Open in Atom
plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse Open in Atom with 0 variables.
plugin:obsidian-shellcommands:188 parseVariables(): Parsing content Open in Atom succeeded.
plugin:obsidian-shellcommands:188 Continuing parsing 'environment_variable_path_augmentation': 
plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse  with 0 variables.
plugin:obsidian-shellcommands:188 parseVariables(): Parsing content  succeeded.
plugin:obsidian-shellcommands:188 Continuing parsing 'stdinContent': undefined
plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse undefined with 0 variables.
plugin:obsidian-shellcommands:188 parseVariables(): Parsing content undefined succeeded.
plugin:obsidian-shellcommands:188 Continuing parsing 'output_wrapper_stdout': undefined
plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse undefined with 0 variables.
plugin:obsidian-shellcommands:188 parseVariables(): Parsing content undefined succeeded.
plugin:obsidian-shellcommands:188 Continuing parsing 'output_wrapper_stderr': undefined
plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse undefined with 0 variables.
plugin:obsidian-shellcommands:188 parseVariables(): Parsing content undefined succeeded.
plugin:obsidian-shellcommands:188 Will call ShellCommandExecutor.executeShellCommand().
plugin:obsidian-shellcommands:188 No augmentation is defined for environment variable PATH. This is completely ok.
plugin:obsidian-shellcommands:188 Executing command open -a Atom \/Users\/ian\/Desktop\/Info\/Projects\/NoiseBin\/Design\/Library\/tools\/log\-install\-czmq\-library\.md in /Users/ian/Desktop/Info/Projects/NoiseBin/Design/Library...
plugin:obsidian-shellcommands:188 No stdin content is present in parsing result.
plugin:obsidian-shellcommands:188 Shell command executed: No errors.```

This should narrow down the location of the issue!

[data.json.txt](https://github.com/Taitava/obsidian-shellcommands/files/10471362/data.json.txt)

**Reply by Taitava** [reply]
Sorry for my absence! I should have mentioned that weekends work better for me for the screenshare. And if needed, I can probably do weekday evenings from 17:00 (UTC+2) / 15:00 (UTC 0) to 20:00 (UTC +2) / 18:00 (UTC 0).

Seems that the log does not reveal the problematic part. Not every step of the program is logged, I've added logging to quite random places. I was hoping that the parsing related logs could reveal something. Well, at least they reveal that the alias text is passed to the parsing process correctly, but it does not show if the parsing process itself makes the alias empty or not, because the `parseVariables(): Parsing content Open in Atom succeeded.` log line does not actually show what the parsing result is - instead it shows the original, unparsed text.

I will improve logging (both for parsing and for the menu opening) and we can agree another screenshare time on Discord. I'll send you a message there.

### Offtopic
The plugin seems to do some useless parsing for `undefined` content and empty content:
> plugin:obsidian-shellcommands:188 Starting to parse 'environment_variable_path_augmentation': 
> plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse  with 41 variables.
> plugin:obsidian-shellcommands:188 parseVariables(): Parsing content  succeeded.
> plugin:obsidian-shellcommands:188 Starting to parse 'stdinContent': undefined
> plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse undefined with 41 variables.
> plugin:obsidian-shellcommands:188 parseVariables(): Parsing content undefined succeeded.
> plugin:obsidian-shellcommands:188 Starting to parse 'output_wrapper_stdout': undefined
> plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse undefined with 41 variables.
> plugin:obsidian-shellcommands:188 parseVariables(): Parsing content undefined succeeded.
> plugin:obsidian-shellcommands:188 Starting to parse 'output_wrapper_stderr': undefined
> plugin:obsidian-shellcommands:188 parseVariables(): Starting to parse undefined with 41 variables.
> plugin:obsidian-shellcommands:188 parseVariables(): Parsing content undefined succeeded.

Those should not affect your issue, but I should make the plugin not to do that kind of unnecessary parsing. This happens when rarely used features (i.e. `stdin` content, `PATH` augmentation and output wrappers) are empty (i.e. not used).

**Reply by Taitava** [reply]
About this older comment of yours:
> Intermittent reflow messages while starting the app or opening ShellCommand config e.g:
> > [Violation] Forced reflow while executing JavaScript took 34ms

Probably something is slow but should not cause this problem, at least if it appears during startup / opening up SC settings, but not during opening the menu.

Offtopic: In case there is dozens of shell commands defined, opening up the shell commands settings has some delay. It's some kind of performance issue, but I haven't looked into it yet.

**Comment by breaker-of-stone** [comment]
See debug logs posted on this issue!

**Comment by Taitava** [comment]
# Summary of the screen share

Thank you @breaker-of-stone for the remote meeting today! I'll write a short summary of my conclusions of the bug just for the record.
- When adding an item to Obsidian's menu, [`Menu.addItem()`](https://github.com/obsidianmd/obsidian-api/blob/38dd22168d2925086371bfc59e36fd9121527a39/obsidian.d.ts#L2256-L2260) is passed a callback function that is responsible for setting up a menu text for the item (and it also sets an icon if wanted).
- The function that the SC plugin passes to Obsidian is [_asynchronous_](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous), meaning that Obsidian's menu creation probably will not wait for the function to finish setting up a text _if it takes long_.
- The function is asynchronous because it needs to parse possible variables in the shell command's alias text in order to show e.g. a file name instead of `{{event_file_name}}` in the menu. Some variables take a long time to parse, so the parsing process has to be asynchronous. @breaker-of-stone's use case does not use variables in the alias text (aka. menu title), so it doesn't even need this parsing, but variable parsing is done anyway just in case.
- The problem is that SC sets the menu text too late. Obsidian doesn't show menu texts that "arrive" late.
- Fix: Set a menu text before parsing any possible variables in it. The text might be e.g. `Open file {{event_file_name}} in app X`. Then update the text after the parsing is done, so it will become e.g. `Open file MyNote.md in app X`. The updated text **will not appear on systems that are affected by this delay problem** (I assume that to be macOS only?). But at least an unparsed text like `Open file {{event_file_name}} in app X` will be shown. It's better than nothing 🙂 .
- For the future, make a feature request to Obsidian that would make [`Menu.addItem()`](https://github.com/obsidianmd/obsidian-api/blob/38dd22168d2925086371bfc59e36fd9121527a39/obsidian.d.ts#L2256-L2260) support asynchronous functions, i.e. so that Obsidian would wait for an asynchronous function to finish before showing the menu.

**Reply by Taitava** [reply]
I'll borrow a Mac from my friend today. Then I'll see if a fix I just wrote will solve the problem.

**Reply by Taitava** [reply]
I was able to confirm the bug on my friend's Mac and also confirm that commit 819fa50777e76f78078fcde9bf32757777f78555 will fix it. I'm trying to release the fix in `0.18.1` today if nothing surprising comes up.

**Reply by Taitava** [reply]
@breaker-of-stone `0.18.1` is now released. Please let me know if you still have any problems or other questions. 🙂

**Comment by breaker-of-stone** [comment]
Understood and accepted.

Thank you once again!
Ian

**Comment by breaker-of-stone** [comment]
Working correctly - awesome!

Thanks Jarkko!

**Reply by Taitava** [reply]
Really nice to hear! 🙂👍🎉
